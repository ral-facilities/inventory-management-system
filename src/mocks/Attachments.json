[
  {
    "entity_id": "1",
    "file_name": "laser-calibration.txt",
    "title": "Laser Calibration",
    "description": "Detailed report on the calibration of high-precision lasers used in experiments.",
    "id": "1",
    "modified_time": "2024-01-02T13:10:10.000+00:00",
    "created_time": "2024-01-01T12:00:00.000+00:00"
  },

  {
    "entity_id": "1",
    "file_name": "safety-protocols.pdf",
    "title": "Safety Protocols",
    "description": "Comprehensive guide outlining safety protocols and procedures for conducting experiments.",
    "id": "1",
    "modified_time": "2024-01-02T13:10:10.000+00:00",
    "created_time": "2024-01-01T12:00:00.000+00:00"
  },

  {
    "entity_id": "1",
    "file_name": "camera-setup-guide.docx",
    "title": "Camera Setup Guide",
    "description": "Guide providing instructions for setting up and configuring cameras used in experiments.",
    "id": "1",
    "modified_time": "2024-01-02T13:10:10.000+00:00",
    "created_time": "2024-01-01T12:00:00.000+00:00"
  },

  {
    "entity_id": "1",
    "file_name": "experiment-results.rtf",
    "title": "Experiment Results",
    "description": "Compilation of results from various experiments conducted, including data analysis and graphs",
    "id": "1",
    "modified_time": "2024-01-02T13:10:10.000+00:00",
    "created_time": "2024-01-01T12:00:00.000+00:00"
  }
]
