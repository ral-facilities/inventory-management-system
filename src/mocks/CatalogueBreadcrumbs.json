[
  {
    "id": "1",
    "trail": [["beam-characterization", "1"]],
    "full_trail": true
  },
  {
    "id": "2",
    "trail": [["motion", "2"]],
    "full_trail": true
  },
  {
    "id": "3",
    "trail": [["vacuum-technology", "3"]],
    "full_trail": true
  },
  {
    "id": "15",
    "trail": [["High-Power-Lasers", "15"]],
    "full_trail": true
  },
  {
    "id": "16",
    "trail": [["X-RAY-Beams", "16"]],
    "full_trail": true
  },
  {
    "id": "4",
    "trail": [
      ["beam-characterization", "1"],
      ["cameras", "4"]
    ],
    "full_trail": true
  },
  {
    "id": "5",
    "trail": [
      ["beam-characterization", "1"],
      ["energy-meters", "5"]
    ],
    "full_trail": true
  },
  {
    "id": "6",
    "trail": [
      ["beam-characterization", "1"],
      ["wavefront-sensors", "6"]
    ],
    "full_trail": true
  },
  {
    "id": "8",
    "trail": [
      ["motion", "2"],
      ["actuators", "8"]
    ],
    "full_trail": true
  },
  {
    "id": "9",
    "trail": [
      ["motion", "2"],
      ["actuators", "8"],
      ["motorized-actuators", "9"]
    ],
    "full_trail": true
  },
  {
    "id": "20",
    "trail": [
      ["motion", "2"],
      ["actuators", "8"],
      ["electric-actuators", "20"]
    ],
    "full_trail": true
  },
  {
    "id": "21",
    "trail": [
      ["motion", "2"],
      ["actuators", "8"],
      ["electric-actuators", "20"],
      ["hydraulic", "21"]
    ],
    "full_trail": true
  },

  {
    "id": "22",
    "trail": [
      ["motion", "2"],
      ["actuators", "8"],
      ["electric-actuators", "20"],
      ["hydraulic", "21"],
      ["linear", "22"]
    ],
    "full_trail": true
  },
  {
    "id": "23",
    "trail": [
      ["actuators", "8"],
      ["electric-actuators", "20"],
      ["hydraulic", "21"],
      ["linear", "22"],
      ["smart", "23"]
    ],
    "full_trail": false
  },

  {
    "id": "10",
    "trail": [
      ["vacuum-technology", "3"],
      ["vacuum-pumps", "10"]
    ],
    "full_trail": true
  },
  {
    "id": "11",
    "trail": [
      ["vacuum-technology", "3"],
      ["vacuum-pumps", "10"],
      ["cryopumps", "11"]
    ],
    "full_trail": true
  },
  {
    "id": "12",
    "trail": [
      ["vacuum-technology", "3"],
      ["vacuum-pumps", "10"],
      ["dry-vacuum-pumps", "12"]
    ],
    "full_trail": false
  },
  {
    "id": "13",
    "trail": [
      ["vacuum-technology", "3"],
      ["vacuum-pumps", "10"],
      ["turbomolecular-pumps", "13"]
    ],
    "full_trail": true
  },
  {
    "id": "14",
    "trail": [
      ["vacuum-technology", "3"],
      ["vacuum-pumps", "10"],
      ["turbo-pumps", "14"]
    ],
    "full_trail": true
  },
  {
    "id": "17",
    "trail": [
      ["High-Power-Lasers", "15"],
      ["Frequency", "17"]
    ],
    "full_trail": true
  },
  {
    "id": "18",
    "trail": [
      ["beam-characterization", "1"],
      ["voltage-meters", "18"]
    ],
    "full_trail": true
  },
  {
    "id": "19",
    "trail": [
      ["beam-characterization", "1"],
      ["amp-meters", "19"]
    ],
    "full_trail": true
  }
]
