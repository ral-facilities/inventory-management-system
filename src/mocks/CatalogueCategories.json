[
  {
    "id": "1",
    "name": "Beam Characterization",
    "parent_id": null,
    "code": "beam-characterization",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "2",
    "name": "Motion",
    "parent_id": null,
    "code": "motion",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "3",
    "name": "Vacuum Technology",
    "parent_id": null,
    "code": "vacuum-technology",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "9086",
    "name": "Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow Overflow",
    "parent_id": null,
    "code": "overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow-overflow",
    "is_leaf": true,
    "properties": [
      {
        "name": "Resolution Overflow Overflow Overflow Overflow Overflow",
        "type": "number",
        "unit": "megapixels",
        "unit_id": "1",
        "mandatory": true,
        "allowed_values": null,
        "id": "78990"
      },
      {
        "name": "Frame Rate",
        "type": "number",
        "unit": "fps",
        "unit_id": "2",
        "mandatory": false,
        "allowed_values": null,
        "id": "78991"
      }
    ],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "79",
    "name": "test_dup",
    "parent_id": "1",
    "code": "test_dup",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "4",
    "name": "Cameras",
    "parent_id": "1",
    "code": "cameras",
    "is_leaf": true,
    "properties": [
      {
        "name": "Resolution",
        "type": "number",
        "unit": "megapixels",
        "unit_id": "1",
        "mandatory": true,
        "allowed_values": null,
        "id": "1"
      },
      {
        "name": "Frame Rate",
        "type": "number",
        "unit": "fps",
        "unit_id": "2",
        "mandatory": false,
        "allowed_values": null,
        "id": "2"
      },
      {
        "name": "Sensor Type",
        "type": "string",
        "unit": null,
        "unit_id": null,
        "mandatory": true,
        "allowed_values": null,
        "id": "3"
      },
      {
        "name": "Sensor brand",
        "type": "string",
        "unit": null,
        "unit_id": null,
        "mandatory": false,
        "allowed_values": null,
        "id": "4"
      },
      {
        "name": "Broken",
        "type": "boolean",
        "unit": null,
        "unit_id": null,
        "mandatory": true,
        "allowed_values": null,
        "id": "5"
      },
      {
        "name": "Older than five years",
        "type": "boolean",
        "unit": null,
        "unit_id": null,
        "mandatory": false,
        "allowed_values": null,
        "id": "6"
      }
    ],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "5",
    "name": "Energy Meters",
    "parent_id": "1",
    "code": "energy-meters",
    "is_leaf": true,
    "properties": [
      {
        "name": "Measurement Range",
        "type": "number",
        "unit": "Joules",
        "unit_id": "3",
        "mandatory": true,
        "allowed_values": null,
        "id": "7"
      },
      {
        "name": "Accuracy",
        "type": "string",
        "unit": null,
        "unit_id": null,
        "mandatory": false,
        "allowed_values": null,
        "id": "8"
      }
    ],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "8967",
    "name": "Energy Meters V2",
    "parent_id": "1",
    "code": "energy-meters-v2",
    "is_leaf": true,
    "properties": [
      {
        "name": "Measurement Range",
        "type": "number",
        "unit": "Joules",
        "unit_id": "3",
        "mandatory": true,
        "allowed_values": null,
        "id": "9"
      },
      {
        "name": "Accuracy",
        "type": "string",
        "unit": null,
        "unit_id": null,
        "mandatory": false,
        "allowed_values": null,
        "id": "10"
      }
    ],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "6",
    "name": "Wavefront Sensors",
    "parent_id": "1",
    "code": "wavefront-sensors",
    "is_leaf": true,
    "properties": [
      {
        "name": "Wavefront Measurement Range",
        "type": "string",
        "unit": null,
        "unit_id": null,
        "mandatory": true,
        "allowed_values": null,
        "id": "11"
      },
      {
        "name": "Spatial Resolution",
        "type": "number",
        "unit": "micrometers",
        "unit_id": "4",
        "mandatory": false,
        "allowed_values": null,
        "id": "12"
      }
    ],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "8",
    "name": "Actuators",
    "parent_id": "2",
    "code": "actuators",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "9",
    "name": "Motorized Actuators",
    "parent_id": "8",
    "code": "motorized-actuators",
    "is_leaf": true,
    "properties": [
      {
        "name": "Travel Range",
        "type": "number",
        "unit": "millimeters",
        "unit_id": "5",
        "mandatory": true,
        "allowed_values": null,
        "id": "13"
      },
      {
        "name": "Load Capacity",
        "type": "number",
        "unit": "kilograms",
        "unit_id": "6",
        "mandatory": true,
        "allowed_values": null,
        "id": "14"
      }
    ],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "20",
    "name": "Electric Actuators",
    "parent_id": "8",
    "code": "electric-actuators",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "21",
    "name": "Hydraulic",
    "parent_id": "20",
    "code": "hydraulic",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "22",
    "name": "Linear",
    "parent_id": "21",
    "code": "linear",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "23",
    "name": "Smart",
    "parent_id": "22",
    "code": "smart",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "10",
    "name": "Vacuum Pumps",
    "parent_id": "3",
    "code": "vacuum-pumps",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "11",
    "name": "Cryo Pumps",
    "parent_id": "10",
    "code": "cryopumps",
    "is_leaf": true,
    "properties": [
      {
        "name": "Cooldown Time",
        "type": "string",
        "unit": null,
        "unit_id": null,
        "mandatory": true,
        "allowed_values": null,
        "id": "15"
      },
      {
        "name": "Pumping Speed",
        "type": "number",
        "unit": "liters per second",
        "unit_id": "7",
        "mandatory": false,
        "allowed_values": null,
        "id": "16"
      }
    ],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "12",
    "name": "Dry Vacuum Pumps",
    "parent_id": "10",
    "code": "dry-vacuum-pumps",
    "is_leaf": true,
    "properties": [
      {
        "name": "Pumping Speed",
        "type": "number",
        "unit": "liters per second",
        "unit_id": "7",
        "mandatory": true,
        "allowed_values": {
          "type": "list",
          "values": [300, 400, 500]
        },
        "id": "17"
      },
      {
        "name": "Ultimate Pressure",
        "type": "number",
        "unit": "millibar",
        "unit_id": "8",
        "mandatory": true,
        "allowed_values": null,
        "id": "18"
      },
      {
        "name": "Axis",
        "type": "string",
        "unit": null,
        "unit_id": null,
        "mandatory": false,
        "allowed_values": {
          "type": "list",
          "values": ["y", "x", "z"]
        },
        "id": "19"
      }
    ],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "13",
    "name": "Turbomolecular Pumps",
    "parent_id": "10",
    "code": "turbomolecular-pumps",
    "is_leaf": true,
    "properties": [
      {
        "name": "Pumping Speed",
        "type": "number",
        "unit": "liters per second",
        "unit_id": "7",
        "mandatory": true,
        "allowed_values": null,
        "id": "20"
      },
      {
        "name": "Compression Ratio",
        "type": "number",
        "mandatory": false,
        "unit": null,
        "unit_id": null,
        "allowed_values": null,
        "id": "21"
      }
    ],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "14",
    "name": "Turbo Pumps",
    "parent_id": "10",
    "code": "turbo-pumps",
    "is_leaf": true,
    "properties": [
      {
        "name": "Pumping Speed",
        "type": "number",
        "unit": "liters per second",
        "unit_id": "7",
        "mandatory": true,
        "allowed_values": null,
        "id": "22"
      },
      {
        "name": "Backing Pressure",
        "type": "number",
        "unit": "millibar",
        "unit_id": "8",
        "mandatory": true,
        "allowed_values": null,
        "id": "23"
      }
    ],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "15",
    "name": "High Power Lasers",
    "parent_id": null,
    "code": "High-Power-Lasers",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "16",
    "name": "X-RAY Beams",
    "parent_id": null,
    "code": "X-RAY-Beams",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "17",
    "name": "Frequency",
    "parent_id": "15",
    "code": "Frequency",
    "is_leaf": true,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "18",
    "name": "Voltage Meters",
    "parent_id": "1",
    "code": "voltage-meters",
    "is_leaf": true,
    "properties": [
      {
        "name": "Measurement Range",
        "type": "number",
        "unit": "volts",
        "unit_id": "9",
        "mandatory": true,
        "allowed_values": null,
        "id": "24"
      },
      {
        "name": "Accuracy",
        "type": "string",
        "unit": null,
        "unit_id": null,
        "mandatory": true,
        "allowed_values": null,
        "id": "25"
      }
    ],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  },
  {
    "id": "19",
    "name": "Amp Meters",
    "parent_id": "1",
    "code": "amp-meters",
    "is_leaf": false,
    "properties": [],
    "created_time": "2024-01-01T12:00:00.000+00:00",
    "modified_time": "2024-01-02T13:10:10.000+00:00"
  }
]
