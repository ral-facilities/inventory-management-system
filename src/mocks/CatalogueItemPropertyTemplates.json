[
  {
    "id": "65c3968fcca15ccd501b0cd0",
    "name": "Material",
    "type": "string",
    "unit": null,
    "mandatory": false,
    "allowed_values": {
      "type": "list",
      "values": ["Fused Silica", "(N)BK-7", "KzFS", "SF6"]
    }
  },
  {
    "id": "65c3968fcca15ccd501b0cd1",
    "name": "Coating",
    "type": "string",
    "unit": null,
    "mandatory": false,
    "allowed_values": {
      "type": "list",
      "values": ["Dielectric", "Protected Gold", "Aluminum", "Enhanced Silver"]
    }
  },
  {
    "id": "65c3968fcca15ccd501b0cd2",
    "name": "Diameter",
    "type": "number",
    "unit": "millimeters",
    "mandatory": true,
    "allowed_values": null
  },
  {
    "id": "65c3968fcca15ccd501b0cd3",
    "name": "Dimension",
    "type": "number",
    "unit": "millimeters",
    "mandatory": true,
    "allowed_values": null
  }
]
