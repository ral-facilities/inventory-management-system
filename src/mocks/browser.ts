import { http, matchRequestUrl } from 'msw';
import { setupWorker } from 'msw/browser';
import { handlers } from './handlers';

// This configures a Service Worker with the given request handlers.

export const worker = setupWorker(...handlers);
// Make the `worker` and `http` references available globally,
// so they can be accessed in both runtime and test suites.
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
window.msw = {
  worker,
  http,
  matchRequestUrl,
};
