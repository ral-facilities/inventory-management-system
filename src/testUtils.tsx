import { LocalizationProvider } from '@mui/x-date-pickers';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFnsV3';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { RenderOptions, render } from '@testing-library/react';
import { enGB } from 'date-fns/locale/en-GB';
import React from 'react';
import {
  RouterProvider,
  createBrowserRouter,
  createMemoryRouter,
} from 'react-router-dom';
import { paths } from './App';
import {
  CatalogueCategory,
  CatalogueCategoryProperty,
  CatalogueItem,
  Item,
  Manufacturer,
} from './api/api.types';
import CatalogueCategoriesJSON from './mocks/CatalogueCategories.json';
import CatalogueItemsJSON from './mocks/CatalogueItems.json';
import ItemsJSON from './mocks/Items.json';
import ManufacturersJSON from './mocks/Manufacturers.json';

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 300000,
      },
    },
  });

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  queryClient?: QueryClient;
}

function constructRouterProvider(
  ui: React.ReactNode,
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const Root: React.FunctionComponent = () => {
    return (
      <LocalizationProvider adapterLocale={enGB} dateAdapter={AdapterDateFns}>
        <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
      </LocalizationProvider>
    );
  };

  const router =
    initialEntry !== undefined && urlPathKey !== undefined
      ? createMemoryRouter([{ path: paths[urlPathKey], Component: Root }], {
          initialEntries: [initialEntry],
        })
      : createBrowserRouter([{ path: '*', Component: Root }]);

  return { router: router, provider: <RouterProvider router={router} /> };
}

function constructRouterProviderWrapper(
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const wrapper = ({
    children,
  }: {
    children: React.ReactNode;
  }): JSX.Element => {
    return constructRouterProvider(
      children,
      queryClient,
      urlPathKey,
      initialEntry
    ).provider;
  };
  return wrapper;
}

export function renderComponentWithRouterProvider(
  ui: React.ReactElement,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string,
  {
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const { router, provider } = constructRouterProvider(
    ui,
    queryClient,
    urlPathKey,
    initialEntry
  );
  return {
    queryClient,
    router,
    ...render(ui, {
      wrapper: () => provider,
      ...renderOptions,
    }),
  };
}

export const hooksWrapperWithProviders = (props?: {
  queryClient?: QueryClient;
  urlPathKey?: keyof typeof paths;
  initialEntry?: string;
}) => {
  const testQueryClient = props?.queryClient ?? createTestQueryClient();
  return constructRouterProviderWrapper(
    testQueryClient,
    props?.urlPathKey,
    props?.initialEntry
  );
};

export const getCatalogueItemsPropertiesById = (
  id: string
): CatalogueCategoryProperty[] => {
  const filteredCategories = CatalogueCategoriesJSON.filter(
    (catalogueCategory) => catalogueCategory.id === id
  );

  if (filteredCategories.length === 0) {
    return [];
  }

  const properties = filteredCategories[0].properties ?? [];
  return properties as CatalogueCategoryProperty[];
};

export const getCatalogueCategoryById = (
  id: string
): CatalogueCategory | undefined => {
  return (
    (CatalogueCategoriesJSON.find(
      (catalogueCategory) => catalogueCategory.id === id
    ) as CatalogueCategory) || undefined
  );
};

export const getCatalogueItemById = (id: string): CatalogueItem | undefined => {
  return (
    (CatalogueItemsJSON.find(
      (catalogueItem) => catalogueItem.id === id
    ) as CatalogueItem) || undefined
  );
};

export const catalogueItemData = (id: string): CatalogueItem[] => {
  return CatalogueItemsJSON.filter(
    (catalogueItem) => catalogueItem.catalogue_category_id === id
  );
};

export const getManufacturerById = (id: string): Manufacturer | undefined => {
  return (
    (ManufacturersJSON.find(
      (manufacturer) => manufacturer.id === id
    ) as Manufacturer) || undefined
  );
};

export const getItemsByCatalogueItemId = (catalogueItemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.catalogue_item_id === catalogueItemId);
};

export const getItemById = (itemId: string): Item | undefined => {
  return ItemsJSON.find((item) => item.id === itemId);
};

export const getItemsBySystemId = (systemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.system_id === systemId);
};

export const CREATED_MODIFIED_TIME_VALUES = {
  created_time: '2024-01-01T12:00:00.000+00:00',
  modified_time: '2024-01-02T13:10:10.000+00:00',
};
