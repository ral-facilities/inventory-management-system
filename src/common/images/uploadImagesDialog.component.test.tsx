import { fireEvent, screen, waitFor } from '@testing-library/react';
import userEvent, { UserEvent } from '@testing-library/user-event';
import { delay, http, HttpResponse } from 'msw';
import { act } from 'react';
import { MockInstance } from 'vitest';
import { server } from '../../mocks/server';
import { renderComponentWithRouterProvider } from '../../testUtils';
import UploadImagesDialog, {
  UploadImagesDialogProps,
} from './uploadImagesDialog.component';

describe('Upload image dialog', () => {
  let props: UploadImagesDialogProps;
  let user: UserEvent;

  let xhrPostSpy: MockInstance;

  const onClose = vi.fn();

  const createView = () => {
    return renderComponentWithRouterProvider(<UploadImagesDialog {...props} />);
  };

  beforeEach(() => {
    props = {
      open: true,
      onClose: onClose,
      entityId: '1',
    };
    user = userEvent.setup();
    xhrPostSpy = vi.spyOn(window.XMLHttpRequest.prototype, 'open');
    global.URL.createObjectURL = vi.fn(() => 'mocked-url');
    vi.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterEach(() => {
    vi.clearAllMocks();
    xhrPostSpy.mockRestore();
  });

  it('renders dialog correctly', async () => {
    let baseElement;
    await act(async () => {
      baseElement = createView().baseElement;
    });

    expect(
      screen.getByText('Files cannot be larger than', { exact: false })
    ).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it('calls onclose when close button is clicked', async () => {
    createView();

    await user.click(screen.getByLabelText('Close Modal'));

    expect(onClose).toHaveBeenCalled();
  });

  it('uploads an image with a title and description and verifies completion message', async () => {
    createView();

    const file1 = new File(['hello world'], 'image.png', {
      type: 'image/png',
    });

    const dropZone = screen.getByText('files cannot be larger than', {
      exact: false,
    });

    Object.defineProperty(dropZone, 'files', {
      value: [file1],
    });

    fireEvent.drop(dropZone, {
      dataTransfer: {
        files: [file1],
      },
    });

    await waitFor(() => {
      expect(screen.getByText('image.png')).toBeInTheDocument();
    });

    await user.click(
      await screen.findByRole('button', { name: 'Edit file image.png' })
    );

    expect(await screen.findByText('File name')).toBeInTheDocument();

    // Checks if file extension is displayed. If it's editable, actual value will disappear after editing.
    expect(screen.getByText('.png')).toBeInTheDocument();

    const name = screen.getByRole('textbox', { name: 'File name .png' });
    const title = screen.getByRole('textbox', { name: 'Title' });
    const description = screen.getByRole('textbox', { name: 'Description' });

    await user.type(name, 'test.jpeg');
    await user.type(title, 'test');
    await user.type(description, 'test');

    expect(screen.getByText('.png')).toBeInTheDocument();

    await user.click(await screen.findByText('Save changes'));

    await user.click(await screen.findByText('Upload 1 file'));

    expect(xhrPostSpy).toHaveBeenCalledWith('POST', '/images', true);

    expect(await screen.findByText('Complete')).toBeInTheDocument();
  });

  it('displays error if post is unsuccessful', async () => {
    server.use(
      http.post('/images', async () => {
        return HttpResponse.error();
      })
    );

    createView();

    const file1 = new File(['hello world'], 'image.png', {
      type: 'image/png',
    });

    const dropZone = screen.getByText('files cannot be larger than', {
      exact: false,
    });

    Object.defineProperty(dropZone, 'files', {
      value: [file1],
    });

    fireEvent.drop(dropZone, {
      dataTransfer: {
        files: [file1],
      },
    });

    await waitFor(() => {
      expect(screen.getByText('image.png')).toBeInTheDocument();
    });

    await user.click(await screen.findByText('Upload 1 file'));

    expect(xhrPostSpy).toHaveBeenCalledWith('POST', '/images', true);

    await waitFor(
      () => {
        expect(screen.getByText('Upload failed')).toBeInTheDocument();
      },
      { timeout: 5000 }
    );
  });

  it(
    'errors when maximum limit is reached',
    async () => {
      server.use(
        http.post('/images', async () => {
          await delay(1000);
          return HttpResponse.json(
            {
              detail:
                'Limit for the maximum number of images for the provided `entity_id` has been reached',
            },
            { status: 422 }
          );
        })
      );

      createView();

      const file1 = new File(['hello world'], 'image.png', {
        type: 'image/png',
      });

      const dropZone = screen.getByText('files cannot be larger than', {
        exact: false,
      });

      Object.defineProperty(dropZone, 'files', {
        value: [file1],
      });

      fireEvent.drop(dropZone, {
        dataTransfer: {
          files: [file1],
        },
      });

      await waitFor(() => {
        expect(screen.getByText('image.png')).toBeInTheDocument();
      });

      await user.click(await screen.findByText('Upload 1 file'));

      expect(xhrPostSpy).toHaveBeenCalledWith('POST', '/images', true);

      await waitFor(
        () => {
          expect(screen.getByText('Upload failed')).toBeInTheDocument();
        },
        { timeout: 10000 }
      );

      await waitFor(() => {
        expect(
          screen.getAllByLabelText('Maximum number of files reached.').length
        ).toBe(2);
      });
    },
    { timeout: 15000 }
  );
});
