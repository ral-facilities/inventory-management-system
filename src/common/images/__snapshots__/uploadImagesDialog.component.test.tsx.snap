// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Upload image dialog > renders dialog correctly 1`] = `
<body
  class="uppy-Dashboard-isFixed"
>
  <div>
    <div
      class="uppy-Container"
    >
      <div
        class="uppy-Root"
        dir="ltr"
      >
        <div
          aria-disabled="false"
          aria-hidden="false"
          aria-label="Uppy Dashboard Window (Press escape to close)"
          class="uppy-Dashboard uppy-Dashboard--modal"
          data-uppy-drag-drop-supported="true"
          data-uppy-num-acquirers="0"
          data-uppy-theme="light"
        >
          <div
            aria-hidden="true"
            class="uppy-Dashboard-overlay"
            tabindex="-1"
          />
          <div
            aria-modal="true"
            class="uppy-Dashboard-inner"
            role="dialog"
          >
            <button
              aria-label="Close Modal"
              class="uppy-u-reset uppy-Dashboard-close"
              title="Close Modal"
              type="button"
            >
              <span
                aria-hidden="true"
              >
                Ã—
              </span>
            </button>
            <div
              class="uppy-Dashboard-innerWrap"
            >
              <div
                class="uppy-Dashboard-dropFilesHereHint"
              >
                Drop your files here
              </div>
              <div
                class="uppy-Dashboard-AddFiles"
              >
                <input
                  accept="image/*"
                  aria-hidden="true"
                  class="uppy-Dashboard-input"
                  hidden=""
                  multiple=""
                  name="files[]"
                  tabindex="-1"
                  type="file"
                />
                <input
                  accept="image/*"
                  aria-hidden="true"
                  class="uppy-Dashboard-input"
                  hidden=""
                  multiple=""
                  name="files[]"
                  tabindex="-1"
                  type="file"
                  webkitdirectory="true"
                />
                <div
                  class="uppy-Dashboard-AddFiles-title"
                >
                  Drop files here or 
                  <button
                    class="uppy-u-reset uppy-c-btn uppy-Dashboard-browse"
                    data-uppy-super-focusable="true"
                    type="button"
                  >
                    browse files
                  </button>
                </div>
                <div
                  class="uppy-Dashboard-AddFiles-list"
                  role="tablist"
                >
                  <span
                    role="presentation"
                    style="white-space: nowrap;"
                  />
                </div>
                <div
                  class="uppy-Dashboard-AddFiles-info"
                >
                  <div
                    class="uppy-Dashboard-note"
                  >
                    Files cannot be larger than 50MB. Only images are allowed.
                  </div>
                </div>
              </div>
              <div
                class="uppy-Dashboard-progressindicators"
              >
                <div
                  aria-hidden="true"
                  class="uppy-StatusBar is-waiting"
                >
                  <div
                    aria-label="0%"
                    aria-valuemax="100"
                    aria-valuemin="0"
                    aria-valuenow="0"
                    aria-valuetext="0%"
                    class="uppy-StatusBar-progress"
                    role="progressbar"
                    style="width: 0%;"
                  />
                  <div
                    class="uppy-StatusBar-actions"
                  />
                </div>
                <div
                  class="uppy uppy-Informer"
                >
                  <span />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
